bucketswatcher_http.go

```diff
@@ -0,0 +1,300 @@
+)
+
+type BucketsWatcherHttpConfig struct {
+	HttpRoundTripper http.RoundTripper
+	Endpoints        []string
+	UserAgent        string
+	Authenticator    Authenticator
+
+	MakeAgent func(ctx context.Context, bucketName string) (*Agent, error)
+}
+
+type BucketsWatcherHttpReconfigureConfig struct {
+	HttpRoundTripper http.RoundTripper
+	Endpoints        []string
+	Authenticator    Authenticator
+}
+
+type BucketsWatcherHttpOptions struct {
+	Logger               *zap.Logger
+	AutoDisconnectPeriod time.Duration
+}
+
+type bucketsWatcherHttpState struct {
+	tlsConfig        *tls.Config
+	authenticator    Authenticator
+	httpRoundTripper http.RoundTripper
+
+	endpoints []string
+}
+
+type BucketsWatcherHttp struct {
+	stateLock sync.Mutex
+	state     *bucketsWatcherHttpState
+	logger    *zap.Logger
+	userAgent string
+	closed    bool
+
+	agents    map[string]*Agent
+	lock      sync.Mutex
+	makeAgent func(ctx context.Context, bucketName string) (*Agent, error)
+
+	watcher        *StreamWatcherHttp[[]bucketDescriptor]
+	watchersCancel func()
+}
+
+func NewBucketsWatcherHttp(cfg BucketsWatcherHttpConfig, opts BucketsWatcherHttpOptions) (*BucketsWatcherHttp, error) {
+	logger := loggerOrNop(opts.Logger)
+
+	watcher, err := NewStreamWatcherHttp[[]bucketDescriptor](&StreamWatcherHttpConfig{
+		HttpRoundTripper: cfg.HttpRoundTripper,
+		Endpoints:        cfg.Endpoints,
+		UserAgent:        cfg.UserAgent,
+		Authenticator:    cfg.Authenticator,
+	}, &StreamWatcherHttpOptions{
+		Logger:               logger.Named("buckets-stream"),
+		AutoDisconnectPeriod: opts.AutoDisconnectPeriod,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	bw := &BucketsWatcherHttp{
+		state: &bucketsWatcherHttpState{
+			authenticator:    cfg.Authenticator,
+			httpRoundTripper: cfg.HttpRoundTripper,
+			endpoints:        cfg.Endpoints,
+		},
+
+		makeAgent: cfg.MakeAgent,
+		userAgent: cfg.UserAgent,
+		logger:    logger,
+		agents:    make(map[string]*Agent),
+		watcher:   watcher,
+	}
+
+	return bw, nil
+}
+
+func (w *BucketsWatcherHttp) Reconfigure(cfg *BucketsWatcherHttpReconfigureConfig) error {
+	w.stateLock.Lock()
+	w.state = &bucketsWatcherHttpState{
+		httpRoundTripper: cfg.HttpRoundTripper,
+		endpoints:        cfg.Endpoints,
+		authenticator:    cfg.Authenticator,
+	}
+	w.stateLock.Unlock()
+
+	w.watcher.Reconfigure(&StreamWatcherHttpReconfigureConfig{
+		HttpRoundTripper: cfg.HttpRoundTripper,
+		Endpoints:        cfg.Endpoints,
+		Authenticator:    cfg.Authenticator,
+	})
+
+	return nil
+}
+
+func (bw *BucketsWatcherHttp) GetAgent(ctx context.Context, bucketName string) (*Agent, error) {
+	bw.lock.Lock()
+	if bw.closed {
+		bw.lock.Unlock()
+		return nil, errors.New("buckets watcher closed")
+	}
+
+	agent, ok := bw.agents[bucketName]
+	if ok {
+		bw.lock.Unlock()
+		return agent, nil
+	}
+	bw.lock.Unlock()
+
+	bw.logger.Debug("Bucket unknown, attempting manual update", zap.String("name", bucketName))
+
+	buckets, err := bw.manuallyFetchBuckets(ctx)
```

Comment:
Risk here that concurrent calls to GetAgent for an unknown bucket will fire off a bunch of requests, but if we want to ensure we're always up to date AFTER the user makes the call then it has to be this way.

Reply:
The best way to solve this I think is that we should record when the user request came in (or at least, record when they invoked `GetAgent`).  Then we deduplicate the manual fetches from ns_server, so that only a single request can be outstanding at once.  When the outstanding request resolves, we resolve any `GetAgent` calls which were before the request was sent.  If there are still pending `GetAgent` calls which need newer information, we can re-dispatch the manual fetch with a newer timestamp which should 'clear the queue' so to speak.